### PR TITLE
Delivering Persistence with Cloudant DB, Config API and basic Fault tolerance support.

### PR DESCRIPTION
**Pull request by jagraj**
Delivering Persistence with Cloudant DB, Config API and basic Fault tolerance support.

### CODE REVIEW

File: manifests/deploy-vote.yaml [hunk 1]
```diff
@@ -1,54 +1,168 @@
...
-  selector:
-    name: vote-deployment
+apiVersion: v1
+kind: Secret
+metadata:
+  name: cloudant-secret
+type: Opaque
+data:
+  dbUsername: YWRtaW4=
+  dbPassword: cGFzcw==
```
Comment (davidcurrie) [inline manifests/deploy-vote.yaml]: Is there some way that we could avoid hard-coding these values? Or at least the password? Generate via a job?

File: manifests/deploy-vote.yaml [hunk 2]
```diff
@@ -1,54 +1,168 @@
...
+---
+kind: PersistentVolume
+apiVersion: v1
+metadata:
+  name: cloudant-pv
+  labels:
+    type: local
+spec:
+  capacity:
+    storage: 4Gi
```
Comment (davidcurrie) [inline manifests/deploy-vote.yaml]: Does it really need 4Gi?!

File: manifests/deploy-vote.yaml [hunk 3]
```diff
@@ -1,54 +1,168 @@
...
+apiVersion: v1
+kind: Service
+metadata:
+  name: cloudant-service
+  namespace: default
+spec:
+  type: NodePort
+  ports:
+    - port: 80
+      nodePort: 31111
```
Comment (davidcurrie) [inline manifests/deploy-vote.yaml]: Do you really need to expose Cloudant as a NodePort? Surely it's just an implementation detail of the service?

File: manifests/deploy-vote.yaml [hunk 4]
```diff
@@ -1,54 +1,168 @@
...
+              key: dbUsername
+        - name: dbPassword
+          valueFrom:
+            secretKeyRef:
+              name: cloudant-secret
+              key: dbPassword
+        command: ["/bin/sh","-c"]
+        args: ["if [ ! -a /srv/cloudant ]; then sed -i 's/172.17.0.2/127.0.0.1/g' /etc/haproxy/haproxy.cfg && cast license --silent && cast database init -v -y -p $dbPassword; fi; exec /usr/sbin/init"]
+        securityContext:
+          privileged: true
```
Comment (davidcurrie) [inline manifests/deploy-vote.yaml]: I'm assuming that this really is required to run Cloudant? It's far from ideal...

File: src/main/java/io/microprofile/showcase/vote/persistence/couch/CouchAttendeeDAO.java [hunk 5]
```diff
@@ -15,50 +15,117 @@
...
+		int maxRetries = 10;
+		Duration duration = Duration.ofSeconds(delayDuration);
+		retryPolicy1 = retryPolicy1.withDelay(duration).withMaxRetries(maxRetries).retryOn(ConnectException.class);
+		// Create an Execution object. Configure it to connect to a "Primary",
+		// with our RetryPolicy
+		// and with a fallback to connect to a "HashMap"
+		Executor executor = FaultToleranceFactory.getFaultToleranceType(Executor.class);
+
+		// Main Service
+		Callable<String> mainService = () -> {
```
Comment (davidcurrie) [inline src/main/java/io/microprofile/showcase/vote/persistence/couch/CouchAttendeeDAO.java]: Seems wrong to have Callable<String> given that we only ever return null.

File: manifests/deploy-vote.yaml (hunk 2 above)
Comment (jagraj) [inline manifests/deploy-vote.yaml]: We have chosen this number if we want to run stress run for this application.

File: manifests/deploy-vote.yaml (hunk 3 above)
Comment (jagraj) [inline manifests/deploy-vote.yaml]: We exposed nodePort to access Cloudant dashboard to monitor data that we are generating for vote service.

File: manifests/deploy-vote.yaml (hunk 4 above)
Comment (jagraj) [inline manifests/deploy-vote.yaml]: Yes, this is workaround for the bug in current cloudant docker image. They are going to fix this problem in next update.

File: src/main/java/io/microprofile/showcase/vote/persistence/couch/CouchAttendeeDAO.java (hunk 5 above)
Comment (jagraj) [inline src/main/java/io/microprofile/showcase/vote/persistence/couch/CouchAttendeeDAO.java]: We have a bug in fault tolerance API, in current beta code we can only return "null" for callable. I opened defect to address this problem. This is workaround until we get an update to beta driver.

**Comment by jagraj** [comment]
We added basic Fault tolerance support for beta, When we deploy vote and cloudant services to the cluster , we need to have cloudant service available for vote service to apply Cloudant DAO objects in post construct. With the retry policy and fallback service we are able to connect to couch DB successfully.  Earlier we had a timing issue  where Cloudant service was not available when vote service is getting started. 
We are planning to enhance vote service for GA release to support HA fail-over when Cloudant service is not available during runtime.

File: manifests/deploy-vote.yaml (hunk 4 above)
Comment (dacleyra) [inline manifests/deploy-vote.yaml]: Both the fix to the internal file, and the privileged security context are required.

**Pull request merged**